cli/defenseclaw/commands/cmd_upgrade.py

```
@@ -706,6 +769,40 @@ def upgrade(
+        # this is the last controller-owned boundary before service stop.
+        try:
+            final_preflight_binding = _read_hard_cut_observability_preflight_binding(
+                data_dir=data_dir,
+                config_path=active_config_path,
+                gateway_binary=gw_binary_path,
+                candidate_directory=staging_dir,
+            )
+            if final_preflight_binding != hard_cut_preflight_binding:
+                raise OSError("hard-cut migration preflight source changed")
+            _require_hard_cut_preflight_state_unchanged(rollback_plan)
+        except BaseException:
```

Ruff (0.15.21)

[warning] Do not catch blind exception: `BaseException`

(BLE001)